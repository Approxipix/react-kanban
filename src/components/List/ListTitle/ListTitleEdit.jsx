import React, { Component } from "react";
import PropTypes from "prop-types";
import { connect } from "react-redux";
import { bindActionCreators } from 'redux';
import { editListTitle } from '../../../redux/listReducer/actions/actions';
import ClickOutside from "../../ClickOutside/ClickOutside";
import { Input } from '../../BaseComponent';
import styled from 'styled-components';

const Form = styled.form`
  width: 90%;
`;

class ListTitleEdit extends Component {
  constructor(props) {
    super(props);
    this.state = {
      title: props.listTitle || '',
    };
  }

  handleChange = (event) => {
    this.setState({
      title: event.target.value,
    });
  };

  handleKeyDown = (e) => {
    if (e.keyCode === 27) {
      this.props.toggleEditTitle()
    }
  };

  handleSubmit = (e) => {
    e.preventDefault();
    const { title } = this.state;
    const { listId } = this.props;
    if (!title)  return;
    this.props.actions.editListTitle({
      listTitle: title,
      listId: listId
    });
    this.props.toggleEditTitle()
  };

  render = () => {
    const { title } = this.state;
    return (
      <ClickOutside toggleOpened={this.props.toggleEditTitle}>
        <Form onSubmit={this.handleSubmit}>
          <Input
            type="text"
            margin="0"
            padding=".3rem .5rem"
            weight="600"
            value={title}
            placeholder="Enter list title..."
            onKeyDown={this.handleKeyDown}
            onChange={(e) => this.handleChange(e)}
            onBlur={this.handleSubmit}
            spellCheck={false}
            autoFocus
          />
        </Form>
      </ClickOutside>
    );
  };
}

ListTitleEdit.defaultProps = {
  listId: '',
  listTitle: '',
  toggleEditTitle: () => {},
};

ListTitleEdit.propTypes = {
  listId: PropTypes.string.isRequired,
  listTitle: PropTypes.string.isRequired,
  toggleEditTitle: PropTypes.func.isRequired,
};

function mapDispatchToProps(dispatch) {
  return {
    actions: bindActionCreators({
      editListTitle: editListTitle,
    }, dispatch)
  };
}

export default connect(null, mapDispatchToProps)(ListTitleEdit);
