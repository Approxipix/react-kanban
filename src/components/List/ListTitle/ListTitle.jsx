import React, { Component } from 'react';
import ListTitleEdit from './ListTitleEdit';
import styled from 'styled-components'

const Title = styled.h2`
  width: 100%;
  padding: .3rem .5rem;
  font-size: 1rem;
  font-weight: 600;
  white-space: nowrap;
  text-overflow: ellipsis;
  overflow: hidden;
`;

class ListHeader extends Component {
  constructor(props) {
    super(props);
    this.state = {
      isEditing: false,
    }
  }

  toggleEditing = () => {
    this.setState({
      isEditing: !this.state.isEditing
    })
  };

  render() {
    const { dragHandleProps, listId, listTitle } = this.props;
    const { isEditing } = this.state;
    return (
      !isEditing ? (
        <Title
          {...dragHandleProps}
          onClick={() => this.toggleEditing()}
        >
          {listTitle}
        </Title>
      ) : (
        <ListTitleEdit
          toggleEditing={this.toggleEditing}
          listId={listId}
          listTitle={listTitle}
        />
      )
    )
  }
}

export default ListHeader;
