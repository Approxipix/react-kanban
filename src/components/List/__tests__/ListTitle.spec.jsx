import React from 'react';
import { shallow } from 'enzyme';
import toJson from "enzyme-to-json";
import ListTitle from '../ListTitle/ListTitle';

describe('<ListTitle>', () => {
  let component;

  beforeEach(() => {
    const state = {
      editTitle: false,
    };

    component = shallow(<ListTitle />);
    component.setState({...state});
  });

  it('should render title if editTitle is false', () => {
    expect(toJson(component)).toMatchSnapshot();
  });

  it('should render ListTitleEdit component if editTitle is true', () => {
    component.instance().toggleEditTitle();
    expect(toJson(component)).toMatchSnapshot();
  });

  it('checks if toggleEditTitle sets editTitle correctly', () => {
    component.instance().toggleEditTitle();
    expect(component.instance().state.editTitle).toEqual(true);

    component.instance().toggleEditTitle();
    expect(component.instance().state.editTitle).toEqual(false);
  });

  it("should call toggleEditTitle after click on title", () => {
    const spy = jest.spyOn(component.instance(), "toggleEditTitle");

    component.find('Title').simulate('click');

    expect(spy).toHaveBeenCalledTimes(1);
  });
});
