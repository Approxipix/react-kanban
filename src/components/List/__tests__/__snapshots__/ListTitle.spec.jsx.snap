// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<ListTitle> should render ListTitleEdit component if editTitle is true 1`] = `
<Connect(ListTitleEdit)
  listId=""
  listTitle=""
  toggleEditTitle={[Function]}
/>
`;

exports[`<ListTitle> should render title if editTitle is false 1`] = `
<Title
  onClick={[Function]}
/>
`;
